File: 138. Copy List with Random Pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        if(head == null) return null;
        HashMap<Node, Node> hm = new HashMap<>();
        Node curr = head;
        while(curr != null){
            hm.put(curr, new Node(curr.val));
            curr = curr.next;
        }

        curr = head;
        Node ans = new Node(curr.val);
        Node ansHead = ans;
        while(curr != null){
            Node temp = hm.get(curr);
            if(!hm.containsKey(curr.random)) temp.random = null;
            else{
                Node randomNode = hm.get(curr.random);
                temp.random = randomNode;
            }
            if(curr.next != null){
                temp.next = hm.get(curr.next);
            }
            curr = curr.next;
        }

        // for(Map.Entry<Node, Node> ele : hm.entrySet()){
        //     Node temp = ele.getValue();
        //     ans.next = temp;
        //     ans = ans.next;
        // }

        return hm.get(head);
    }
}

File: 138. Copy List with Random Pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if (!head) return nullptr;

    // Step 1: Create new nodes and insert them next to the original nodes
        Node* current = head;
        while (current) {
            Node* newNode = new Node(current->val);
            newNode->next = current->next;
            current->next = newNode;
            current = newNode->next;
        }
        current = head;
        while (current) {
            if (current->random) {
                current->next->random = current->random->next;
            }
            current = current->next->next;
        }

        // Step 3: Separate the newly created list from the original list
        current = head;
        Node* newHead = head->next;
        Node* copy = newHead;
        while (current) {
            current->next = current->next->next;
            if (copy->next) {
                copy->next = copy->next->next;
            }
            current = current->next;
            copy = copy->next;
        }

        return newHead;
    }
};